stustapay/core/service/order.py

```diff
@@ -65,90 +93,115 @@ async def create_order(
 
-        if status != "pending":
-            raise Exception("order not in pending state")
+        prepared_bookings: Dict[Tuple[int, int, str], float] = defaultdict(lambda: 0.0)
+        for line_item in order.line_items:
+            product = line_item.product
+            source_account_id = get_source_account(ORDER_TYPE_SALE, product, customer.id)
+            target_account_id = get_target_account(ORDER_TYPE_SALE, product, customer.id)
+            # only positive bookings are allowed, swap accounts if price is negative
+            if product.price < 0:
+                source_account_id, target_account_id = target_account_id, source_account_id
+            prepared_bookings[(source_account_id, target_account_id, line_item.tax_name)] += line_item.total_price
+
+        # execute the created bookings
+        for (source_account_id, target_account_id, tax_name), amount in prepared_bookings.items():
+            await conn.fetchval(
+                "select * from book_transaction("
+                "   order_id => $1,"
+                "   description => $2,"
+                "   source_account_id => $3,"
+                "   target_account_id => $4,"
+                "   amount => $5,"
+                "   tax_name => $6)",
+                order.id,
+                "",
+                source_account_id,
+                target_account_id,
+                amount,
+                tax_name,
+            )
 
-        # TODO create transactions accordingly and call 'select book_transaction()' in the database
-        # sufficent funds are not checked in the database
-        # current available funds of source
+        return await conn.fetchval("select balance from account where id = $1", customer.id)
 
+    async def _finish_order(self, *, conn: asyncpg.Connection, order: Order):
+        """
+        Once the order is executed properly, mark it as done and notify tse or bon service to process this order
+        """
         # mark the order as done
         await conn.fetchval(
-            "update ordr set finished_at = now(),  status = 'done' , customer_account_id = $2 where id = $1;",
-            order_id,
-            customer_account_id,
+            "update ordr set finished_at = now(),  status = 'done' where id = $1;",
+            order.id,
         )
 
         # TODO first add a TSE signing request
         # The TSE signer should then call the code below to add the bon request
         # create bon request
         await conn.fetchval(
-            "insert into bon(id) values($1) "
-            "on conflict do"
-            "    update set generated = false, "
-            "               generated_at = null, "
-            "               status = null "
-            "    where id = $1;",
-            order_id,
+            "insert into bon(id) values($1) ",
+            order.id,
+        )
+        await conn.execute("select pg_notify('bon', $1);", str(order.id))
```

Comment:
out of curiosity: is this sent when the dbtransaction is committed, or immediately?

Reply:
When it is commited, of course. See the [Postgres Doku](https://www.postgresql.org/docs/current/sql-notify.html)
> NOTIFY interacts with SQL transactions in some important ways. Firstly, if a NOTIFY is executed inside a transaction, the notify events are not delivered until and unless the transaction is committed.